app/coffee/ResourceWriter.coffee

```diff
@@ -11,7 +11,39 @@ settings = require("settings-sharelatex")
 parallelFileDownloads = settings.parallelFileDownloads or 1
 
 module.exports = ResourceWriter =
-	syncResourcesToDisk: (project_id, resources, basePath, callback = (error) ->) ->
+
+	syncResourcesToDisk: (request, basePath, callback = (error) ->) ->
+		if request.incremental?
+			ResourceWriter.checkState request.incremental, basePath, (error, ok) ->
+				logger.log state: request.state, result:ok, "checked state on incremental request"
+				return callback new Error("invalid state") if not ok
+				ResourceWriter.saveIncrementalResourcesToDisk request.project_id, request.resources, basePath, callback
+		else
+			@saveAllResourcesToDisk request.project_id, request.resources, basePath, (error) ->
+				return callback(error) if error?
+				ResourceWriter.storeState request.state, basePath, callback
+
+	storeState: (state, basePath, callback) ->
+		logger.log state:state, basePath:basePath, "writing state"
+		fs.writeFile Path.join(basePath, ".resource-state"), state, {encoding: 'ascii'}, callback
+
+	checkState: (state, basePath, callback) ->
+		fs.readFile Path.join(basePath, ".resource-state"), {encoding:'ascii'}, (err, oldState) ->
+			logger.log state:state, oldState: oldState, basePath:basePath, err:err, "checking state"
+			if state is oldState
+				return callback(null, true)
+			else
+				return callback(null, false)
+
+	saveIncrementalResourcesToDisk: (project_id, resources, basePath, callback = (error) ->) ->
```

Comment:
Hmm, you're right, the logic of OutputFileFinder doesn't work here.  From web we either need to be sending in a list of all the files in the request, with some flagged as 'unchanged', or keep that list in the .project-sync-state file.